### PR TITLE
Canvas 42 music reserves backend

### PR DESCRIPTION
**Pull request by maxpluez**

### CODE REVIEW

File: src/server/services/UpdateReserveServices.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-class UpdateVideoResServices {
+class UpdateReserveServices {
```
Comment (rlorenzo) [inline src/server/services/UpdateReserveServices.js:1]: @maxpluez I see now both Video Reserves and Music Reserves is sharing the same code.. can this also work for BruinCast?

File: src/server/jobs/update-musicres.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      await UpdateMusicResServices.updateRecordsForClass(
+        dbclient,
+        'musicreserves',
+        term,
+        srs,
+        works,
+        logger
+      );
+    }
+    logger.info('done');
```
Comment (rlorenzo) [inline src/server/jobs/update-musicres.js:65]: @maxpluez Can you include a report of how many records were added, like in BruinCast?

https://github.com/ucla/lti-media-resources/blob/master/src/server/jobs/update-bcast.js#L290

**Comment by maxpluez** [comment]
This new push is just a fix on count feature. Fixing other problems now..

File: migrations/20200804193836-CreateCollectionsIfNotExist.js [hunk 3]
```diff
@@ -1,18 +1,21 @@
...
     if (!collectionList.map(col => col.name).includes('crosslists')) {
       await db.createCollection('crosslists');
     }
     if (!collectionList.map(col => col.name).includes('bruincastmedia')) {
       await db.createCollection('bruincastmedia');
     }
+    if (!collectionList.map(col => col.name).includes('videoreserves')) {
+      await db.createCollection('videoreserves');
+    }
+    if (!collectionList.map(col => col.name).includes('musicreserves')) {
```
Comment (rlorenzo) [inline migrations/20200804193836-CreateCollectionsIfNotExist.js:13]: @maxpluez Please add musicreserves to a new migration file.

File: src/server/services/MediaResourceServices.js [hunk 4]
```diff
@@ -17,10 +17,14 @@ class MediaResourceServices {
       courseLabel
     );
 
+    const musicresCount = await MediaQuery.getMusicResCountByCourse(
+      courseLabel
```
Comment (rlorenzo) [inline src/server/services/MediaResourceServices.js:21]: @maxpluez So it seems that once someone is authenticated via LTI they can then, potentially, call our end points for any course label and get the media counts. That seems like a security concern. The endpoint already knows the context.label the user is logged into. Can't we get that data from the LTIjs ContextToken (https://cvmcosta.me/ltijs/#/provider?id=contexttoken)? Why have it passed as a parameter?

File: src/server/models/mediaquery.js [hunk 5]
```diff
@@ -43,6 +43,19 @@ module.exports.getVideoResCountByCourse = async courseLabel => {
...
+module.exports.getMusicResByCourse = async courseLabel => {
+  const musicResCollection = client.db(DB_DATABASE).collection('musicreserves');
+  const toBeReturned = await musicResCollection
+    .find({ classShortname: courseLabel })
+    .toArray();
+  return toBeReturned;
+};
+
+module.exports.getMusicResCountByCourse = async courseLabel => {
+  const arrayOfMusicRes = await this.getMusicResByCourse(courseLabel);
```
Comment (rlorenzo) [inline src/server/models/mediaquery.js:55]: @maxpluez For all these end points, you should be ensuring the user has a role and belongs to that given label. Or get the label from the ContextToken.

File: src/server/services/MediaResourceServices.js (hunk 4 above)
Comment (maxpluez) [reply src/server/services/MediaResourceServices.js:21]: Wait, this function is only called by the api routes. the api routes call this function with context token.

Comment (rlorenzo) [reply src/server/services/MediaResourceServices.js:21]: @maxpluez  what line is that call happening?

Comment (rlorenzo) [reply src/server/services/MediaResourceServices.js:21]: @maxpluez  okay I see it at src/server/api/medias/musicres.js

Still should check role

Comment (maxpluez) [reply src/server/services/MediaResourceServices.js:21]: It's in server/api/medias/index.js

Comment (maxpluez) [reply src/server/services/MediaResourceServices.js:21]: Oh yeah both of them call it

Comment (maxpluez) [reply src/server/services/MediaResourceServices.js:21]: So technically speaking all js files under /services are only in service to the api routes. Users shouldn't be able to directly call them..

File: src/server/models/mediaquery.js (hunk 5 above)
Comment (maxpluez) [reply src/server/models/mediaquery.js:55]: Added role check under api route /api/media/index.js, which calls this function.

File: src/server/api/medias/bruincast.js [hunk 6]
```diff
@@ -32,6 +32,17 @@ router.post('/crosslists', (req, res) => {
...
+  const roles = res.locals.token.roles.map(role =>
+    role.substr(role.lastIndexOf('#') + 1, role.length).toLowerCase()
+  );
+  if (
+    !roles.includes('learner') &&
+    !roles.includes('teacher') &&
+    !roles.includes('instructor') &&
+    !roles.includes('administrator')
+  ) {
+    return res.status(400);
```
Comment (rlorenzo) [inline src/server/api/medias/bruincast.js:44]: @maxpluez Error 400 is Bad Request. It makes more sense to return 403, which is Forbidden

File: src/server/api/medias/musicres.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
+const express = require('express');
+
+const router = express.Router();
+
+const MusicresServices = require('../../services/MusicresServices');
+
+router.get('/', (req, res) => {
+  const roles = res.locals.token.roles.map(role =>
```
Comment (rlorenzo) [inline src/server/api/medias/musicres.js:8]: @maxpluez This is violating the DRY principle: https://en.wikipedia.org/wiki/Don%27t_repeat_yourself

All this repeated code means there is need for a function. See @kilometerskang made a file to contain this logic: https://github.com/ucla/lti-library-resources/blob/70c423ef086fd6cf0ac96c9f5af288f1b6c79b75/src/client/components/App/userRoles.js#L3

File: src/server/services/MusicresServices.js [hunk 8]
```diff
@@ -0,0 +1,10 @@
+const MediaQuery = require('../models/mediaquery');
+
+class MusicresServices {
+  static async getMusicres(label) {
+    const docs = await MediaQuery.getMusicResByCourse(label);
+    return docs.sort((a, b) => a.title - b.title);
```
Comment (rlorenzo) [inline src/server/services/MusicresServices.js:6]: @maxpluez Isn't it more efficient to let the database sort the results? https://stackoverflow.com/questions/1883264/database-sort-vs-programmatic-java-sort

**Pull request merged**